#New calculation of Exac Exon-level constraint.
#This script is written so that it may call code from other files
#But will not require any preprocessing of data (as much as possible)
#3/19/2016
#NWK
library(dplyr)
library(lazyeval)
#SetOps.R contains helper functions for performing intersection, union, etc. on genomic intervals.
#SetOps.R requires the installation of bedops (https://bedops.readthedocs.io/en/latest/)
#SetOps.R also requires annovar (http://annovar.openbioinformatics.org/en/latest/user-guide/download/)
source("~/Dropbox/AnnotateR/SetOps.R")
#ExacFuncs has helper functions for parsing the Exac VCF data.  #ExacFuncs requires vcftools
#(https://vcftools.github.io/index.html)
source("~/Dropbox/BayesianDA/Exac/ExacFuncs.R")


#This is the working directory
hdir <- "~/Desktop/Exac/temp/"
#This is the directory that contains the annovar data
dbdir = "/media/nwknoblauch/Data/annovar/humandb/"
#This is the directory where Exac Data is kept
ExacDataDir <- "~/Desktop/Exac/temp/"
#This is the directory where results will be stored
ResultDir <- "~/Dropbox/BayesianDA/Exac/"
#This is the directory containing coverage data
covdir <- "/media/nwknoblauch/Data/ExacCoverage/"
#This is the path to the gene level Exac Z-scores (obtained from Xin, but a similar file can be obtained from 
#ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3/functional_gene_constraint/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt)
#I've created a dropbox folder with many of the necessary files here:
#https://www.dropbox.com/sh/9njd385qb8eotwo/AAC8OkYtS9h9j27hh8YT6pkOa?dl=0
genef <- file.path(hdir,"Kaitlin_esp6500_ac10_Zdata_fordist.csv")
synonf <- file.path(hdir,"MutRate_exomeWide.synonymous.txt")
nsynonf <- file.path(hdir,"MutRate_exomeWide.nonsynonymous.txt")
#This is the compressed exac VCF file.  A copy can be found on PPS
# (/mnt/gluster/home/nwknoblauch/Downloads/ExAC.r0.3.sites.vep.vcf.gz)
Exac_All_VCF <- file.path(hdir,"ExAC.r0.3.sites.vep.vcf.gz")
Exac_All_of <- file.path(hdir,"ExAC_All")
Exac_All_info <- paste0(Exac_All_of,".INFO")


#These lines allow easy switching between all Exac cases and the non-psychiatric cases
Exac_VCF <- Exac_All_VCF
Exac_of <- Exac_All_of
Exac_info <- Exac_All_info


#This generates the R friendly csv with allele count an quality score for every site
VCFFunc(Exac_VCF,gzipped = T,fields = c("AC_Adj","AN_Adj","VQSLOD"),outfile = Exac_All_of)
#VCFFunc(Exac_NP_VCF,gzipped = T,fields = c("AC_Adj","AN_Adj","VQSLOD"),outfile = Exac_NP_of)


#If you are trying to generate exon-level annotations, set RegionClass to "Exon", otherwise set it to something else
#This section reads in synonymous and non-synonymous mutation rates and merges them 
#(If we're computing exon-level z-scores, otherwise it uses the pre-computed gene-level Z scores)
RegionClass <- "Exon"
if(RegionClass=="Exon"){
  synondf <- read.table(synonf,header=F,stringsAsFactors = F)
  nsynondf <- read.table(nsynonf,header=F,stringsAsFactors = F)
  colnames(synondf) <- c("chr","start","end","gene","syn_rate")
  colnames(nsynondf) <- c("chr","start","end","gene","nsyn_rate")
  region_df <- full_join(synondf,nsynondf)
}else{
  genedf <- read.table(genef,header=T,stringsAsFactors = F,sep="\t")
  sgenedf <- select(genedf,chr,start,end=stop,gene,syn_rate=p_syn,nsyn_rate=p_mis) %>% mutate(chr=paste0("chr",chr))
  region_df  <- sgenedf
}

#Here we convert the info file generated by vcftools (called in VCFFunc) into a bed file for use with bedops 
Exac_bed_full <- info.bed(Exac_info) %>% mutate(af_adj=as.numeric(ac_adj)/as.numeric(an_adj))
#This command filters out the common variants (you might want to get rid of this, depending on what you're doing)
Exac_bedd <-filter(Exac_bed_full,af_adj<0.001)
#This prepares and saves the reference Exac data as an RDS file
Exac_rdsf <- file.path(ExacDataDir,paste0("Reference_Exac_full_df.RDS"))
saveRDS(Exac_bed_full,file = Exac_rdsf)

#We need coverage information for the gene level or the exon level, depending on the type of annotation we're 
#interested in.
ref_f <- file.path(ExacDataDir,paste0("Reference_Exac_",RegionClass,".txt"))
covmf = file.path(covdir,paste0(RegionClass,"_covmap.txt"))

#Find the exon assignment that has no overlap and covers the largest portion of the genome
#(I fully acknowledge how awkward this line is, and will try to find time to  refactor it in the future)
region_df <-  semi_join(region_df,
                        rename(
                          FindMax(select(bdf,chrom=region_df,start,end,name=gene))
                          ,gene=name)) %>%
  mutate(start=format(start,scientific=F,trim=T),end=format(end,scientific=F,trim=T))
#Write the exon (or gene) table to a file
write.table(region_df,ref_f,col.names=T,sep="\t",row.names=F,quote=F)
#This function computes region level coverage (it takes a really long time to run)
#The necessary coverage data can be downloaded here 
#ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3/coverage
cov_df <- ExacCov(covdir = covdir,exonf = ref_f,covmf = covmf)
Exac_varf <- annovar_bed(bdf = Exac_bed,dbdir=dbdir,tmpdir = ExacDataDir ,name = VariantClass)

Exac_synvarf <- file.path(ExacDataDir,paste0(VariantClass,"_synon.txt"))
Exac_nsynvarf <- file.path(ExacDataDir,paste0(VariantClass,"_nonsynon.txt"))

#Split the exac variants into synonymous and non-synonymous 
splitdf(Exac_varf,Exac_synvarf,Exac_nsynvarf)


#Count the types of each class of variant
Exac_cvf <- varcount(Exac_synvarf,Exac_nsynvarf,ref_f,cov_exdf = cov_df)
Exac_ref_ctf <- file.path(ResultDir,paste0(VariantClass,"_Reference_Exac_",RegionClass,".txt"))

#Write the counts to a file
write.table(Exac_cvf,Exac_ref_ctf,col.names=T,row.names = F,quote=F)


#These are high-leverage outliers in the linear model(of mutation rate and base to mutation count) that I manually found, and remove 
exclude.exons <- data_frame(start=c(42976313,
                                    105404399,
                                    100633911,
                                    103381800,
                                    103381800,
                                    144990344,
                                    9056172),
                            gene=c("STARD9",
                                   "AHNAK2",
                                   "MUC12",
                                   "CCDC168",
                                   "CCDC168",
                                   "PLEC",
                                   "MUC16"))

exclude.gene=c("TTN")

#This function calculates the exac z score, excluding the data points above
#This could also use some refactoring
compute_exac_z <- function(df,include.sex=FALSE,exclude.gene=c("TTN"),exlude.exons=NULL){

  if(!include.sex){
    ndf <- filter(df,chrom %in% paste0("chr",1:22))
  }else{
    ndf <- df
  }
  ndf <- filter(ndf,!gene %in% exclude.gene)
  ndf <- filter(ndf,n_syn!=0,n_syn!=0)
  ndf <- mutate(ndf,bp=end-start)
  if(!is.null(exclude.exons)){
    ndf <- anti_join(ndf,exclude.exons)
  }
  ndf <- filter(ndf,!is.na(syn_rate),!is.na(nsyn_rate))


  ndf <- group_by(ndf,gene) %>% mutate(rn=paste0(gene,":",1:n())) %>% ungroup()
  rownames(ndf) <- ndf$rn
  tndf <- filter(ndf,median>50)
  syn.pred.lm.bp <- lm(n_syn~syn_rate+bp,data=tndf)
  ndf$pred_syn <- predict(syn.pred.lm.bp,newdata=ndf)
  ndf <- mutate(ndf,pred_syn)
  ndf$pred_nsyn <- predict(syn.pred.lm.bp,newdata = data_frame(syn_rate=ndf$nsyn_rate,bp=ndf$bp))
  ndf <- mutate(ndf,med_cut_bin=cut(median,breaks=c(0,seq(to=max(median),by=2.0),max(median)+1),labels = F,include.lowest = T))
  ndf <- mutate(ndf,med_cut_bin=(med_cut_bin/max(med_cut_bin,na.rm = T))*max(median))
  #From Samocha et al method (slightly different values from the above fit)
  #Also see page 69 of supplement
  ndf <- mutate(ndf,adj_p_syn=ifelse(median>=50,pred_syn,pred_syn*(0.09189+0.27086*log(pmax(1,median)))))
  ndf <- mutate(ndf,adj_p_nsyn=ifelse(median>=50,pred_nsyn,pred_nsyn*(0.09189+0.27086*log(pmax(1,median)))))
  ndf <- mutate(ndf,adj_p_nsyn=pmax(1,adj_p_nsyn))
  ndf <- mutate(ndf,mis_cs=(n_nsyn-adj_p_nsyn)^2/adj_p_nsyn,
                z_sign_mis=sqrt(mis_cs)*ifelse(n_nsyn>adj_p_nsyn,-1,1))
  ndf <- select(ndf,-med_cut_bin,-mis_cs)
  return(ndf)

}

Exac_X <- filter(NP_nexondf,chrom=="chrX")
#Run on autosomes
Exac_final_auto <- compute_exac_z(df=Exac_cvf,   include.sex = F,exclude.gene=c("TTN"),exlude.exons = exclude.exons)
#Run on the X chromosme
Exac_final_X <-    compute_exac_z(df=Exac_X,include.sex = T,exclude.gene=c("TTN"),exlude.exons = NULL)
#combine both
Exac_final <- bind_rows(Exac_final_auto,NP_final_X)
#Write results to csv
write.table(Exac_final,file.path(ResultDir,paste0(VariantClass,"Exac_",RegionClass,"_Cons.txt")),
            col.names=T,row.names=F,sep="\t",quote=F)
#Write results to RDS
saveRDS(final_exon,file.path(ResultDir,paste0(VariantClass,"Exac_",RegionClass,"_Cons.RDS")))


